lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php

```diff
@@ -104,11 +106,32 @@ public function getDefinitions()
      * @param array $args
      *
      * @return object
-     *
+     * @throws \Exception
      */
     protected function createObject($type, $args)
     {
-        return new $type(...array_values($args));
+        try {
+            return new $type(...array_values($args));
+        } catch (\Throwable $e) {
+            $this->getLogger()->critical(__(
+                'Object create error: %1, %2',
+                [
+                    $type,
+                    $e->getMessage()
+                ]
+            ));
+            throw new RuntimeException(__($e->getMessage()));
+        }
+    }
+
+    /**
+     * Logger Instance
+     *
+     * @return LoggerInterface
+     */
+    protected function getLogger()
+    {
+        return $this->objectManager->get(LoggerInterface::class);
```

Comment:
Classes extending this abstraction (for example `\Magento\Framework\ObjectManager\Factory\Compiled`) would not have `objectManager` available. See constructor for this class and descendants.

I'd still suggest covering this with tests, as multiple undesired effects may be caused by this type of recursion. Basically, now object manager's `createObject` calls for object manager once more in a specific case, which may end up in an endless loop.